### PR TITLE
refactor backend

### LINKED ISSUE #10: add native s3 storage backend
I was having a look at implementing it myself, but found the backend traits to not be async.  I also found in the changelog, that async was removed in 0.4.  Is this going to change soonish again?  If so, I'll defer it.

I would use `rust-s3`, because afaict the official aws sdk is async only.  What do you think?

### LINKED ISSUE #11: No sftp support?
Hi!

Love the project, thank you so much! When I tried to use my existing restic repository (Hetzner Storagebox), rustic lets me know that the sftp backend is not supported. Am I missing something? :)

Thank you :D

### PR DESCRIPTION
**Pull request by aawsome**
Besides refactoring the backend traits and allowing to use self-made backends, this PR also includes the new opendal backend which allows access to various cloud storages including SFTP, S3, B2, GCS, Openstack Swift, and many more...

## BREAKING CHANGES

- The `Repository::ls()` method now returnes an iterator with yielding elements with restricted lifetime. You may need to change your code to handle this, see e.g. https://github.com/rustic-rs/rustic/commit/5be54a1bfe4e6c63299e343c905c28be45860269

- config changes from `[repository]` to `[backend]` (check `config/`)
- for rustic core API: the way how to construct a repository changed, `struct RepositoryOptions` changed
- check also examples for changes

fixes #10
fixes #11 
closes #26 

Todo:
- [x] decision about separator between provider string and url in repository string for less ambiguity, use `|` or `>` or `;` instead of `:`? -> stick to `:`
- [X] feedback for fake repo and dealing (https://github.com/rustic-rs/rustic/blob/d674904f5f6805c3c8b340a5eb6351f4e17fedad/src/commands/key.rs#L65)
-> see `RepositoryOptions::evaluate_password` 
- [X] parse to url/path in url_to_type_and_path(), for better future error handling (e.g. on Windows) -> sensible for UI, do later, open issue
- [ ] any other breaking changes to be made?
- [x] discuss release process for `rustic_core` and `rustic_backend`: just release rustic_core and rustic_backend, if we do it timely it should be fine
- [ ] discuss config changes and make them consistent across the ecosystem (scheduler config)

Open points:
- [x] error handling
- [x] prepare extracting backend implementations to separate crate
- [x] Add opendal backend
- [x] check doc-comments

### CODE REVIEW

**Comment by aawsome** [comment]
> @aawsome First feedback: Could you add a `BREAKING CHANGES:` abstract to the changelog in this PR please that documents what breaking changes have been made for users of the API and also WRT the new backends?

I checked the changes and actually found just one breaking change, see the PR description. BTW: Of course there are also tons of breaking changes in things which are declared `pub` (e.g. in `error.rs`) - but those are currently not available to rustic_core users as they (and the corresponding modules) are not exported in `lib.rs`.

File: src/backend/opendal.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+            url.set_query(None);
+            url.set_fragment(None);
+            let _ = options.insert("endpoint".to_string(), url.to_string());
+        }
+        _ = options
+            .entry("region".to_string())
+            .or_insert_with(|| "auto".to_string());
+
+        Self::new("s3", dbg!(options))
+    }
```
[the remark this answers is not in the document]
Comment (aawsome) [reply src/backend/opendal.rs:69]: Definitively. I would do both: Allow all `opendal` services by using `opendal:<SERVICE>` as repo url - which needs to set additional options in the `[repository.options]` section and allows to directly add new services without any rustic_core change by simply updating opendal.

And I think we should also allow to use something like `s3:https://<HOST>/<BUCKET>/<ROOT>` and so on for services where the corresponding opendal service is "mature" enough such that we can be sure we want to use and propagate it. For S3 I think this is the case as the service for instance also allows to use the standard S3 env variables or aws profile, IIRC.

File: src/backend.rs [hunk 2]
```diff
@@ -215,13 +233,11 @@ pub trait ReadBackend: Clone + Send + Sync + 'static {
     ///
     /// # Errors
     ///
-    /// * [`IdErrorKind::HexError`] - If the string is not a valid hexadecimal string
-    /// * [`BackendErrorKind::NoSuitableIdFound`] - If no id could be found.
-    /// * [`BackendErrorKind::IdNotUnique`] - If the id is not unique.
+    /// * [`BackendAccessErrorKind::NoSuitableIdFound`] - If no id could be found.
+    /// * [`BackendAccessErrorKind::IdNotUnique`] - If the id is not unique.
     ///
-    /// [`IdErrorKind::HexError`]: crate::error::IdErrorKind::HexError
```
[the remark this answers is not in the document]
Comment (aawsome) [reply src/backend.rs:222]: No, it is also used in other places, e.g. in `src/repofile/snapshotfile.rs`.

File: crates/backend/src/util.rs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl TryFrom<&str> for BackendLocation {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &str) -> std::result::Result<Self, Self::Error> {
+        match (canonicalize(Path::new(value)), Url::parse(value), value) {
```
Comment (aawsome) [inline crates/backend/src/util.rs:23]: canonicalize returns an error if the path doesn't exist. We can't use this in combination with `init` - here we actually want to create a path, but will run into `Invalid backend location`....

**Review by aawsome — COMMENTED** [review]
LGTM

**Comment by devhell** [comment]
This is AMAZING! Thank you! 🍻

**Pull request merged**